### PR TITLE
feat: implement RFC-7

### PR DESCRIPTION
**Pull request by menduz**

### CODE REVIEW

File: scripts/generate-proto-file.ts [hunk 1]
```diff
@@ -143,14 +144,17 @@ proto.addMessage(
...
   "StreamMessage",
   [
     PORT_ID,
     field("fixed32", "sequence_id", 4), // sequence id of the StreamMessage, starting at 0
     PAYLOAD, // Server->Client the payload in case of a successful result
     field("bool", "closed", 7), // Server->Client/Client->Server an empty message signaling the end of the stream
     field("bool", "ack", 8), // Client->Server used for backpressure
+    field("bool", "require_ack", 9), // Server->Client used to know if the stream will use `ack`
+    field("bool", "client_stream", 10), // Open client stream
+    field("bool", "server_stream", 11), // Open server stream
```
Comment (menduz) [inline scripts/generate-proto-file.ts:157]:
```suggestion
```

File: src/client.ts [hunk 2]
```diff
@@ -92,117 +91,64 @@ function throwIfRemoteError(parsedMessage: RemoteError) {
...
+            sendStreamThroughTransport(
+              requestDispatcher.dispatcher,
+              requestDispatcher.dispatcher.transport,
+              data as any,
+              portId,
+              messageNumber
+            )
+          })
+          .catch((error) => {
+            debugger
```
Comment (menduz) [inline src/client.ts:140]:
```suggestion
```

File: src/codegen.ts [hunk 3]
```diff
@@ -30,10 +32,35 @@ export function clientProcedureUnary<Request, Response>(
   return fn
 }
 
-export function clientProcedureStream<Request, Response>(
+async function* requestToBinaryGenerator<Request, Response>(requests: AsyncIterable<Request>, method: TsProtoMethodDefinition<Request, Response>): AsyncIterable<Uint8Array> {
+  for await (const request of requests) {
+    const arg = method.requestType.encode(request)
+    yield arg.finish()
+  }
+  // console.log('codegen: requestToBinaryGenerator done!')
```
Comment (menduz) [inline src/codegen.ts:40]:
```suggestion
```

File: src/message-dispatcher.ts [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  })
+
+  function receiveAck(data: StreamMessage, messageNumber: number) {
+    const key = `${messageNumber},${data.sequenceId}`
+    const fut = ackCallbacks.get(key)
+    if (fut) {
+      ackCallbacks.delete(key)
+      fut[0](data)
+      // } else {
+      //   throw new Error("Received an ACK message for an inexistent handler " + key)
```
Comment (menduz) [inline src/message-dispatcher.ts:96]:
```suggestion
      } else {
        throw new Error("Received an ACK message for an inexistent handler " + key)
```

File: src/push-channel.ts [hunk 5]
```diff
@@ -186,6 +186,7 @@ export class AsyncQueue<T> implements AsyncGenerator<T> {
   constructor(private requestingNext: (queue: AsyncQueue<T>, action: "next" | "close") => void) { }
 
   [Symbol.asyncIterator](): AsyncGenerator<T> {
+    debugger
```
Comment (menduz) [inline src/push-channel.ts:189]:
```suggestion
```

File: src/server.ts [hunk 6]
```diff
@@ -399,7 +402,11 @@ export function createRpcServer<Context = {}>(options: CreateRpcServerOptions<Co
       messageType == RpcMessageTypes.RpcMessageTypes_STREAM_ACK ||
       messageType == RpcMessageTypes.RpcMessageTypes_STREAM_MESSAGE
     ) {
-      ackHelper.receiveAck(parsedMessage, messageNumber)
+      // } else if (messageType == RpcMessageTypes.RpcMessageTypes_STREAM_ACK) {
+      //   dispatcher.receiveAck(parsedMessage, messageNumber)
+      // } else if (messageType == RpcMessageTypes.RpcMessageTypes_STREAM_MESSAGE) {
+      //   const receivedStreamMessage = parsedMessage as StreamMessage
+      //   dispatcher.emitStream(messageNumber, receivedStreamMessage)
```
Comment (menduz) [inline src/server.ts:409]:
```suggestion
```

File: test/client-stream.spec.ts [hunk 7]
```diff
@@ -0,0 +1,5 @@
+describe("client stream", () => {
+  it("a client stream tests", () => {
+    console.log("hello world")
+  })
+})
```
Comment (menduz) [inline test/client-stream.spec.ts:5]: delete file

File: src/message-dispatcher.ts (hunk 4 above)
Comment (kuruk-mm) [reply src/message-dispatcher.ts]: https://github.com/decentraland/rpc/issues/116

**Pull request merged**